libs/storage/Tsavorite/cs/src/core/Index/Recovery/Recovery.cs

```diff
@@ -735,14 +787,20 @@ private void RecoverHybridLogFromSnapshotFile(long scanFromAddress, long recover
         {
             GetSnapshotPageRangesToRead(scanFromAddress, untilAddress, snapshotStartAddress, snapshotEndAddress, guid, out long startPage, out long endPage, out long snapshotEndPage, out int capacity, out var recoveryStatus, out int numPagesToReadFirst);
 
-            hlog.AsyncReadPagesFromDevice(startPage, numPagesToReadFirst, snapshotEndAddress,
-                                          hlog.AsyncReadPagesCallbackForRecovery,
-                                          recoveryStatus, recoveryStatus.recoveryDevicePageOffset,
-                                          recoveryStatus.recoveryDevice, recoveryStatus.objectLogRecoveryDevice);
-
-            for (long page = startPage; page < endPage; page += capacity)
+            for (long page = startPage; page < endPage; page += numPagesToReadFirst)
             {
-                long end = Math.Min(page + capacity, endPage);
+                var readEndPage = Math.Min(page + numPagesToReadFirst, snapshotEndPage);
+                if (page < readEndPage) // Read pages max up to snapshotEndPage
+                {
+                    var numPagesToReadThisIteration = (int)(readEndPage - page);
+
+                    // Ensure that page slots that will be read into, have been flushed from previous reads. 
+                    WaitUntilAllPagesHaveBeenFlushed(page, readEndPage, recoveryStatus);
+
+                    ReadPagesWithMemoryConstraint(snapshotEndAddress, capacity, recoveryStatus, page, readEndPage, numPagesToReadThisIteration);
```

Comment:
Consolidate the body of this block into a function rather than repeating the lines here, in RecoverHybridLog, and RecoverHybridLogAsync, and RecoverHybridLogFromSnapshotFileAsync